<div align='center'><picture><h1><b>Data Analysis Process</b></h1></picture></div>

# **Article 130 : Data Assessing & Cleaning** [![Static Badge](https://img.shields.io/badge/Open%20in%20Colab%20-%20orange?style=plastic&logo=googlecolab&labelColor=grey)](https://colab.research.google.com/github/sshrizvi/DataScienceMastery/blob/main/DataAnalysisProcess/Notebooks/130_data_assessing_and_cleaning.ipynb)

|🔴 **NOTE** 🔴|
|:-----------:|
| This notebook contains the practical implementations of the concepts discussed in the following article.|
| Here is Article 130 - [Data Assessing & Cleaning](../Articles/130_data_assessing_and_cleaning.md) |

As of now, we know that *Data Analysis Process* is a variable process that do not have a certain (*or fixed*) framework, through which we can complete the analysis on our data. Many authors have given various frameworks to perform analysis. This framework can also depend on the domain of data we have to deal with (*like airlines, transport, disease, schools, etc*).  

For this notebook we are following this framework:  
1. Asking Questions
2. Data Wrangling
   1. Data Gathering [(Learn)](./129_data_gathering.ipynb)
   2. Data Assessing (*In this Notebook*)
   3. Data Cleaning (*In this Notebook*)
3. Exploratory Data Analysis
4. Drawing Conclusion
5. Communicating Results

### 🎯 **Data Assessing**

In the Data Assessing Step, we will have to understand the data that we have gathered in the previous step of Data Analysis Process.  
It requires you to understand the description of :  
1. Domain of the Data
2. Variables Involved
3. Context
You will have to study the data in as much detail as possible. It is going to help us identify the problems in the data, that will ultimately help in the Data Cleaning.

### 🎯 **Unclean Data**

There are majorly two types of unclean data that we can think of:  
1. **Dirty Data**  
   Dirty Data, is also known as *Low Quality Data* (or Data with Quality Issues). This kind of data has content issues like:
      1. Duplicated Data
      2. Inaccurate Data
      3. Missing Data
      4. Corrupt Data  
   
   For Example, A table contains the data of physical properties (height, weight, color) of some students.  
   Some of the values in the `height` column, are negative values, which is irrational (or inaccurate).

2. **Messy Data**  
   Messy Data, also known as *Untidy Data* (or Data with Tidiness Issues). This kind of data has structural issues. Tidy Data has the following properties:
      1. Each variable forms a column.
      2. Each observation forms a row.
      3. Each obervational unit forms a table.  
   

### 📦 **Importing Relevant Libraries**

In [159]:
import pandas as pd
import numpy as np

### ⚠️ **Data Warning**
Now lets import an unknown dataset on which we can perform data assessing and cleaning.  
Here is the link to the dataset : [Link](https://github.com/campusx-official/data-wrangling)  
You can also find these files in the [Data](../Resources//Data/) Folder.

In [160]:
adverse_rxn_df = pd.read_csv('../Resources/Data/adverse_reactions.csv')

In [161]:
patients_df = pd.read_csv('../Resources/Data/patients.csv')

In [162]:
treatments_df = pd.read_csv('../Resources/Data/treatments.csv')

In [163]:
treatments_cut_df = pd.read_csv('../Resources/Data/treatments_cut.csv')

Once your data is imported, the first thing that you will have to find out the context (*or summary*) of the data.  
The summary provides you with an overview of the dataset, and helps you to assess the dataset in an accessible and accurate manner.  
It is possible that this summary is already provided to you by the data provider, in case you downloaded it from some kind of data repository, like `Kaggle`. In some cases, you might have to write the summary of the data by youself. An example of this case would be like:  
- The dataset that you have is provided by your organisation. In that case, you can ask your manager or concerned authority about the summary.
- The dataset that you have is scraped by you from some source. In that case, you will have to write the summary by yourself.

Now, lets proceed with writing the summary about our dataset, that is, collection of the following four DataFrames (*or Tables*):  
1. adverse_rxn_df
2. patients_df
3. treatments_df
4. treatments_cut_df

### 🎯 **Summary of *Auralin* Drug Trials Dataset**

This is a dataset about *500* patients of which *350* patients participated in a *clinical trial*. None of the patients were using **Novodra** (a popular injectable insulin) or **Auralin** (the oral insulin being researched) as their primary source of insulin before. All were experiencing elevated `HbA1c` levels.

All *350* patients were treated with **Novodra** to establish a baseline `HbA1c` level and insulin dose.  
After 4 weeks, which isn’t enough time to capture all the change in `HbA1c` that can be attributed by the switch to **Auralin** or **Novodra**:

- ***175*** patients switched to **Auralin** for *24* weeks
- ***175*** patients continued using **Novodra** for *24* weeks

**Note :** Data about patients feeling some adverse effects is also recorded.

Once you have the summary of the dataset, you will have to understand the variables used in the dataset, which represents the values or feature that has been recorded for each student.  
So, lets start with writing the description about the variables involved in the dataset.  
Since, we have `four` tables related to the dataset, we will go one by one.

### 💊 **Table *Patients***

| No. | Field Name | Description |
|----:|-----------|-------------|
| 1 | patient_id | Unique identifier for each patient in the Master Patient Index (patient database) of the pharmaceutical company producing Auralin |
| 2 | assigned_sex | Assigned sex of the patient at birth (male or female) |
| 3 | given_name | Given name (first name) of the patient |
| 4 | surname | Surname (last name) of the patient |
| 5 | address | Main residential address of the patient |
| 6 | city | City corresponding to the patient’s main address |
| 7 | state | State corresponding to the patient’s main address |
| 8 | zip_code | Zip code corresponding to the patient’s main address |
| 9 | country | Country corresponding to the patient’s main address (United States for this clinical trial) |
| 10 | contact | Phone number and email information for the patient |
| 11 | birthdate | 1. Date of birth of the patient (MM/DD/YYYY). <br> 2. Inclusion criteria: age ≥ 18 years |
| 12 | weight | Weight of the patient in pounds (lbs) |
| 13 | height | Height of the patient in inches (in) |
| 14 | bmi | 1. Body Mass Index (BMI) of the patient  <br> 2. Calculated as: **BMI = kg/m²**  <br> 3. Healthy range: **18.5–24.9**  <br> 4. Inclusion criteria: **16 ≤ BMI ≤ 38** |


### 💊 **Table *Treatments***

| No. | Field Name | Description |
|----:|-----------|-------------|
| 1 | given_name | 1. Given name of each patient <br> 2. Patient is part of the Master Patient Index <br> 3. Patient participated in the clinical trial |
| 2 | surname | 1. Surname (last name) of each patient <br> 2. Patient is part of the Master Patient Index <br> 3. Patient participated in the clinical trial |
| 3 | auralin | 1. Baseline median daily insulin dose from the week prior to switching to **Auralin** (value before the dash) <br> 2. Ending median daily insulin dose at the end of **24 weeks** of treatment (value after the dash) <br> 3. Dose measured during the **24th week of treatment** <br> 4. Unit of measurement: **units (u)** — international standard for insulin |
| 4 | novodra | 1. Same structure as **auralin** <br> 2. Applies to patients who continued treatment with **Novodra** <br> 3. Baseline median daily insulin dose before treatment period <br> 4. Ending median daily insulin dose at the end of 24 weeks <br> 5. Unit of measurement: **units (u)** |
| 5 | hba1c_start | 1. Patient’s **HbA1c level** at the beginning of the first week of treatment <br> 2. HbA1c (Hemoglobin A1c) reflects average blood sugar over the past **3 months** <br> 3. Used to assess overall diabetes control <br> 4. Measured in **percentage (%)** |
| 6 | hba1c_end | 1. Patient’s **HbA1c level** at the end of the last week of treatment <br> 2. Measured in **percentage (%)** |
| 7 | hba1c_change | 1. Change in HbA1c from start to end of treatment <br> 2. Calculated as: **hba1c_start − hba1c_end** <br> 3. Used to assess treatment effectiveness <br> 4. **Noninferiority criterion:** <br> &nbsp;&nbsp;4.1 Upper bound of 95% CI < 0.4% <br> &nbsp;&nbsp;4.2 Comparison: **Novodra − Auralin** <br> &nbsp;&nbsp;4.3 Must meet threshold for Auralin to be deemed effective |


### 💊 **Table *Treatments Cut***

**Note :** The columns of `treatments` and `treatments_cut` table are same. The reason behind this is that, out of *350* peoples who participated in the clinical trial of **Auralin**, data about the treatments or doses is collected only for *280* peoples in the `treatments` table. Remaining *70* peoples data is in the `treatments_cut` table.  


| No. | Field Name | Description |
|----:|-----------|-------------|
| 1 | given_name | 1. Given name of each patient <br> 2. Patient is part of the Master Patient Index <br> 3. Patient participated in the clinical trial |
| 2 | surname | 1. Surname (last name) of each patient <br> 2. Patient is part of the Master Patient Index <br> 3. Patient participated in the clinical trial |
| 3 | auralin | 1. Baseline median daily insulin dose from the week prior to switching to **Auralin** (value before the dash) <br> 2. Ending median daily insulin dose at the end of **24 weeks** of treatment (value after the dash) <br> 3. Dose measured during the **24th week of treatment** <br> 4. Unit of measurement: **units (u)** — international standard for insulin |
| 4 | novodra | 1. Same structure as **auralin** <br> 2. Applies to patients who continued treatment with **Novodra** <br> 3. Baseline median daily insulin dose before treatment period <br> 4. Ending median daily insulin dose at the end of 24 weeks <br> 5. Unit of measurement: **units (u)** |
| 5 | hba1c_start | 1. Patient’s **HbA1c level** at the beginning of the first week of treatment <br> 2. HbA1c (Hemoglobin A1c) reflects average blood sugar over the past **3 months** <br> 3. Used to assess overall diabetes control <br> 4. Measured in **percentage (%)** |
| 6 | hba1c_end | 1. Patient’s **HbA1c level** at the end of the last week of treatment <br> 2. Measured in **percentage (%)** |
| 7 | hba1c_change | 1. Change in HbA1c from start to end of treatment <br> 2. Calculated as: **hba1c_start − hba1c_end** <br> 3. Used to assess treatment effectiveness <br> 4. **Noninferiority criterion:** <br> &nbsp;&nbsp;4.1 Upper bound of 95% CI < 0.4% <br> &nbsp;&nbsp;4.2 Comparison: **Novodra − Auralin** <br> &nbsp;&nbsp;4.3 Must meet threshold for Auralin to be deemed effective |


### 💊 **Table *Adverse Reactions***

| No. | Field Name | Description |
|----:|-----------|-------------|
| 1 | given_name | 1. Given name of each patient <br> 2. Patient is part of the Master Patient Index <br> 3. Patient participated in the clinical trial <br> 4. Patient experienced an adverse reaction <br> 5. Includes patients treated with **Auralin** and **Novodra** |
| 2 | surname | 1. Surname (last name) of each patient <br> 2. Patient is part of the Master Patient Index <br> 3. Patient participated in the clinical trial <br> 4. Patient experienced an adverse reaction <br> 5. Includes patients treated with **Auralin** and **Novodra** |
| 3 | adverse_reaction | 1. Adverse reaction reported by the patient <br> 2. Reaction occurred during participation in the clinical trial <br> 3. May vary in severity and type |


Now, we are ready with the summary of our dataset. As of now, we have the following things:  
1. Summary of our *Auralin* Drug Trials Dataset.
2. Detailed Description about the Variables involved in each of the tables involved.

It feels like we have all the details about our dataset, but still there are some extra information that might be present there.  
If there are certain other things available, there is no harm in gathering and documenting those. It is gonna help us in ways.

### 👽 **Extra Information**
1. Insulin resistance varies person to person, which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
2. It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This diversity is reflected in the patients table.

### 🎯 **Assessment Types**

We can assess our data in two ways:  
1. ***Manual*** : It involves looking through the data manually in softwares like *Google Sheets* or *Microsoft Excel*.
2. ***Programmatic*** : It involves using Pandas methods like `info()`, `describe()`, `sample()`, etc.

### 🎯 **Steps in Assessment**
The process of assessment, whether *manual* or *programmatically*, involves two steps:
1. ***Discover*** : In this step, you go through your data, and find the inaccuracies in it.
2. ***Document*** : In this step, you write the identified inaccuracies that you found in the previous discovery step.

To assess our data, we are going to export it into an Excel file. This will help us in discovering our complete data.

In [164]:
output_path = '../Resources/Exports/auralin_clinical_trials.xlsx'
with pd.ExcelWriter(output_path) as writer:
    patients_df.to_excel(writer,sheet_name='patients',
                         index=False)
    treatments_df.to_excel(writer, sheet_name='treatments',
                           index=False)
    treatments_cut_df.to_excel(writer, sheet_name='treatments_cut',
                               index=False)
    adverse_rxn_df.to_excel(writer, sheet_name='adverse_reactions',
                            index=False)

### 🎯 **Issues in Dataset**

First, we will be starting to look for issues (*both dirty & messy*) in the dataset.  
While looking manually for the issues in the dataset, I will be documenting it here, classifying as *dirty* & *messy*.

1. **Dirty Issues**
   1. ***Table : Patients***
      1. **Manual**
         1. The value of `given_name` for `patient_id = 9` is *Dsvid* instead of *David*.
         2. There are some *missing values* in columns `[address, city, state, zip_code, country, contact]`.
         3. In the `state` column, values are inconsistent. Some values are *abbreviations*, whereas some are *full forms*.
         4. In the `zip_code` column, values are inconsistent. Some values are *4-digit* zip code, which is not possible.
      2. **Programmatically**
         1. There are **12** *missing values* in columns `[address, city, state, zip_code, country, contact]`.
         2. Column `assigned_sex` should have `pd.Categorical` as its data type.
         3. Column `zip_code` should have `int64` as its data type.
         4. Column `birthdate` should have `pd.DateTime` as its data type.
         5. Duplicated Entries by the name **John Doe** with different `patient_id` and some other duplicate entries by following names:  
            1. **Jakob Jakobsen** and **Jake Jakobsen** are same.
            2. **Patrick Gersten** and **Pat Gersten** are same.
            3. **Sandra Taylor** and **Sandy Taylor** are same.  
         6. Unusual value of `weight` & `height` were seen in `patient_id` 5 & 211.
   2. ***Table : Treatments & Treatments Cut***
      1. **Manual** 
         1. The `given_name` and `surname` has values in lower case, which is not consistent with the *Patients Table*.
         2. Remove `u` from the values of `auralin` and `novodra`, since the units of all values are same.
         3. Hyphens (-) in `auralin` and `novodra` columns, are actually `nan` values, but ain't getting treated like that.
         4. Column `hba1c_change` contains some *missing values*.
      2. **Programmatically**
         1. Column `hba1c_change` contains **109** *missing values* in `treatments_df` and **28** in `treatments_cut_df`.
         2. While splitting `auralin`, assign proper data types to `auralin_start` and `auralin_end`.
         3. While splitting `novodra`, assign proper data types to `novodra_start` and `novodra_end`.
         4. A duplicate entry exists by the name **joseph day** in `treatments_df`.
         5. **56** values in `hba1c_change` column are miscalculated.
   3. ***Table : Adverse Reactions***
      1. The `given_name` and `surname` has values in lower case, which is not consistent with the *Patients Table*.
2. **Messy Issues**
   1. ***Table : Patients***
      1. The `contact` column contains both *phone number* & *email*.
   2. ***Table : Treatments & Treatments Cut***
      1. **Manual**
         1. The column `auralin` should get splitted into two columns `auralin_start` and `auralin_end`.
         2. The column `novodra` should get splitted into two columns `novodra_start` and `novodra_end`.
         3. Combine `treatments` and `treatments_cut` tables into one.
   3. ***Table : Adverse Reactions***
      1. This table is capturing information about one variable only, so it should not exist independently.

### 🤖 **Programmatic Assessment**

In the programmatic assessment we will be using inbuilt Pandas Methods to analyse our data and issues hidden in it. Also, we can use a little bit of logic to find the problems in the dataset.  
So, lets first start with using Pandas Methods.

#### **1. info Method**

The `info` Method of Pandas provides the following information about a DataFrame:  
1. Total no. of rows, and information about the index.
2. Count of non-null values in each of the columns of DataFrame.  
   *This can help in identifying null values in the DataFrame*
3. Data Type of each column.  
   *This can help in identifying the data type of columns in the DataFrame*

In [165]:
patients_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


Now, as we can see that in `patients_df`, we have some `null` values in the columns  
```python
[address, city, state, zip_code, country, contact]
```  
Also, it is visible that some of the columns do not have proper data types assigned. Here is a list of those:  
1. Column `assigned_sex` should have `pd.Categorical()` as its data type.
2. Column `zip_code` is assigned with data type `float64`, whereas the proper data type would be `int64`.
3. Column `birthdate` should have `DateTime` as Data Type.

So, lets document it above, in the issues part of this notebook.

In [166]:
treatments_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


Now, as we can see that in the `treatments_df`, we have **109** `null` values in `hba1c_change` column.  
Also the data type of columns `auralin` and `novodra` should be assigned properly while splitting both of them, as observed during manual assessment. So, lets document this too.

In [167]:
treatments_cut_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


Now, as we can see that in the `treatments_cut_df`, we have **28** `null` values in `hba1c_change` column.  
Also the data type of columns `auralin` and `novodra` should be assigned properly while splitting both of them, as observed during manual assessment. So, lets document this too.

In [168]:
adverse_rxn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


Now, as we can see that in `adverse_rxn_df`, we do not have any problems in terms of `null` values and data types.

#### **2. isnull Method**

The `isnull` method tells us about the null entries in our DataFrame. Since it returns a `boolean` mask, where `True` represents `null` value and `False` represents `non-null` values, we can actually fetch those rows from our DataFrame to assess manually.

In [169]:
patients_df.isnull()

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
0,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
498,False,False,False,False,False,False,False,False,False,False,False,False,False,False
499,False,False,False,False,False,False,False,False,False,False,False,False,False,False
500,False,False,False,False,False,False,False,False,False,False,False,False,False,False
501,False,False,False,False,False,False,False,False,False,False,False,False,False,False


In [170]:
mask = patients_df['address'].isnull()
patients_df[mask]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
209,210,female,Lalita,Eldarkhanov,NaN,NaN,NaN,NaN,NaN,NaN,8/14/1950,143.4,62,26.2
219,220,male,Mỹ,Quynh,NaN,NaN,NaN,NaN,NaN,NaN,4/9/1978,237.8,69,35.1
230,231,female,Elisabeth,Knudsen,NaN,NaN,NaN,NaN,NaN,NaN,9/23/1976,165.9,63,29.4
234,235,female,Martina,Tománková,NaN,NaN,NaN,NaN,NaN,NaN,4/7/1936,199.5,65,33.2
242,243,male,John,O'Brian,NaN,NaN,NaN,NaN,NaN,NaN,2/25/1957,205.3,74,26.4
249,250,male,Benjamin,Mehler,NaN,NaN,NaN,NaN,NaN,NaN,10/30/1951,146.5,69,21.6
257,258,male,Jin,Kung,NaN,NaN,NaN,NaN,NaN,NaN,5/17/1995,231.7,69,34.2
264,265,female,Wafiyyah,Asfour,NaN,NaN,NaN,NaN,NaN,NaN,11/3/1989,158.6,63,28.1
269,270,female,Flavia,Fiorentino,NaN,NaN,NaN,NaN,NaN,NaN,10/9/1937,175.2,61,33.1
278,279,female,Generosa,Cabán,NaN,NaN,NaN,NaN,NaN,NaN,12/16/1962,124.3,69,18.4


**Observation :** It is clear that these rows are those where values in columns  
```python
['address', 'city', 'state', 'zip_code', 'country', 'contact']
```
are missing and are exactly **12**, as observed during `info` method.

#### **3. duplicated Method**

The `duplicated` method of Pandas tells us about the rows or columns in the DataFrame that are duplicated.

In [171]:
patients_df.duplicated().sum()

0

So, its clear that there is no row duplicated in the DataFrame.  
But did you thought about how `duplicated` method works?  
The answer is that, when I ran `patients_df.duplicated().sum()`. It attempts to find out that how many rows are there in the DataFrame for which values of each column of those rows matches. For instance :  
```python
[1, female, Zoe, Wellish, 576 Brown Bear Drive, Rancho California, California, 92390.0, United States, 951-719-9170ZoeWellish@superrito.com, 7/10/1976, 121.7 ,66 ,19.6]
[1, female, Zoe, Wellish, 576 Brown Bear Drive, Rancho California, California, 92390.0, United States, 951-719-9170ZoeWellish@superrito.com, 7/10/1976, 121.7 ,66 ,19.6]
```
are duplicated, as values of each column is matching. Whereas,  
```python
[1, female, Zoe, Wellish, 576 Brown Bear Drive, Rancho California, California, 92390.0, United States, 951-719-9170ZoeWellish@superrito.com, 7/10/1976, 121.7 ,66 ,19.6]
[1, female, Zoe, Wellish, 2370 University Hill Road, Armstrong, Illinois, 61812.0, United States, PamelaSHill@cuvox.de+1 (217) 569-3204, 4/3/1967, 118.8, 66, 19.2]
```
are not duplicated, as values of each column is not matching. But as you can see in the above code snippet that both rows contains the inforamtion about the same person `Zoe Wellish`, which is not possible.  

Now, to tackle such kind of search using `duplicated` method, we will use the `subset` parameter of this method. All we have to provide is combination of column names.

Now, lets try to check that if `patients_df` contains any duplicate entries by same name, that is, `[given_name, surname]`.

In [172]:
mask = patients_df.duplicated(subset=['given_name', 'surname'])
patients_df[mask]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
229,230,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
237,238,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
244,245,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
251,252,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
277,278,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4


**Observation :** It can be seen that there are **5** duplicated entries of a person **John Doe**, with almost all information being same. It was just `patient_id` that was stopping these from getting identifed without `subset` parameter.  
So, lets document it in issues.

Now, lets check if there are any entries that have same addresses.

In [173]:
mask = (patients_df.duplicated(subset=['address'], keep=False)) & (patients_df['address'].notnull())
patients_df[mask]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
24,25,male,Jakob,Jakobsen,648 Old Dear Lane,Port Jervis,New York,12771.0,United States,JakobCJakobsen@einrot.com+1 (845) 858-7707,8/1/1985,155.8,67,24.4
29,30,male,Jake,Jakobsen,648 Old Dear Lane,Port Jervis,New York,12771.0,United States,JakobCJakobsen@einrot.com+1 (845) 858-7707,8/1/1985,155.8,67,24.4
97,98,male,Patrick,Gersten,2778 North Avenue,Burr,NE,68324.0,United States,PatrickGersten@rhyta.com402-848-4923,5/3/1954,138.2,71,19.3
131,132,female,Sandra,Taylor,2476 Fulton Street,Rainelle,WV,25962.0,United States,304-438-2648SandraCTaylor@dayrep.com,10/23/1960,206.1,64,35.4
215,216,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
229,230,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
237,238,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
244,245,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
251,252,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
277,278,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4


**Observation :** With this trick of `address`, we get to know that apart from **John Doe**, we have three more peoples for whom the data is duplicated. Here are those persons.  
1. **Jakob Jakobsen** and **Jake Jakobsen** are same.
2. **Patrick Gersten** and **Pat Gersten** are same.
3. **Sandra Taylor** and **Sandy Taylor** are same.  
This information is verified by manually assessing the other details of these persons.  

So, lets document it in issues.

Now, lets check for duplicacy in `contact` column.

In [174]:
mask = (patients_df.duplicated(subset=['contact'], keep=False)) & (patients_df['contact'].notnull())
patients_df[mask]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
24,25,male,Jakob,Jakobsen,648 Old Dear Lane,Port Jervis,New York,12771.0,United States,JakobCJakobsen@einrot.com+1 (845) 858-7707,8/1/1985,155.8,67,24.4
29,30,male,Jake,Jakobsen,648 Old Dear Lane,Port Jervis,New York,12771.0,United States,JakobCJakobsen@einrot.com+1 (845) 858-7707,8/1/1985,155.8,67,24.4
97,98,male,Patrick,Gersten,2778 North Avenue,Burr,NE,68324.0,United States,PatrickGersten@rhyta.com402-848-4923,5/3/1954,138.2,71,19.3
131,132,female,Sandra,Taylor,2476 Fulton Street,Rainelle,WV,25962.0,United States,304-438-2648SandraCTaylor@dayrep.com,10/23/1960,206.1,64,35.4
215,216,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
229,230,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
237,238,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
244,245,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
251,252,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
277,278,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4


**Observation :** No issues found. The values are same as seen earlier.

Now, lets check for duplicacy in `birthdate` column.

In [175]:
mask = (patients_df.duplicated(subset=['birthdate'], keep=False)) & (patients_df['birthdate'].notnull())
patients_df[mask].sort_values(by='birthdate')

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
215,216,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
229,230,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
237,238,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
244,245,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
251,252,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
277,278,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
131,132,female,Sandra,Taylor,2476 Fulton Street,Rainelle,WV,25962.0,United States,304-438-2648SandraCTaylor@dayrep.com,10/23/1960,206.1,64,35.4
282,283,female,Sandy,Taylor,2476 Fulton Street,Rainelle,WV,25962.0,United States,304-438-2648SandraCTaylor@dayrep.com,10/23/1960,206.1,64,35.4
204,205,male,Leixandre,Alanis,4103 Musgrave Street,Oklahoma City,OK,73102.0,United States,LeixandreAlanisMadrigal@fleckens.hu405-234-5350,3/25/1938,222.6,69,32.9
480,481,male,Nasser,Mansour,547 Weekley Street,San Antonio,TX,78212.0,United States,NasserMazinMansour@fleckens.hu1 210 326 5509,3/25/1938,183.5,66,29.6


**Observation :** No issues found. Though some duplicate `birthdate` are found, it is possible that many persons were born on same day. It is verified by manual assessment.

Now, lets check for duplicacy for the combination `[address, zip_code]`.

In [176]:
mask = (patients_df.duplicated(subset=['address', 'zip_code'], keep=False)) & (patients_df['zip_code'].notnull())
patients_df[mask].sort_values(by='zip_code')

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
215,216,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
229,230,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
237,238,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
244,245,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
251,252,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
277,278,male,John,Doe,123 Main Street,New York,NY,12345.0,United States,johndoe@email.com1234567890,1/1/1975,180.0,72,24.4
24,25,male,Jakob,Jakobsen,648 Old Dear Lane,Port Jervis,New York,12771.0,United States,JakobCJakobsen@einrot.com+1 (845) 858-7707,8/1/1985,155.8,67,24.4
29,30,male,Jake,Jakobsen,648 Old Dear Lane,Port Jervis,New York,12771.0,United States,JakobCJakobsen@einrot.com+1 (845) 858-7707,8/1/1985,155.8,67,24.4
131,132,female,Sandra,Taylor,2476 Fulton Street,Rainelle,WV,25962.0,United States,304-438-2648SandraCTaylor@dayrep.com,10/23/1960,206.1,64,35.4
282,283,female,Sandy,Taylor,2476 Fulton Street,Rainelle,WV,25962.0,United States,304-438-2648SandraCTaylor@dayrep.com,10/23/1960,206.1,64,35.4


**Observation :** Same values again. No issue found.

Now, lets analyze `treatments_df` and `treatments_cut_df` using `duplicated` method.

In [177]:
mask = treatments_df.duplicated(keep=False)
treatments_df[mask]

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
10,joseph,day,29u - 36u,-,7.7,7.19,NaN
136,joseph,day,29u - 36u,-,7.7,7.19,NaN


**Observation :** A duplicated entry exists in `treatments_df` by the name of **joseph day**. So, lets document it.

Now, lets use some different combinations of columns to assess duplicacy issues.

In [178]:
mask = treatments_df.duplicated(subset=['given_name', 'surname'], keep=False)
treatments_df[mask]

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
10,joseph,day,29u - 36u,-,7.7,7.19,NaN
136,joseph,day,29u - 36u,-,7.7,7.19,NaN


**Observation :** Same values again. No issues found.

Now, lets analyze `adverse_rxn_df` for duplicacy issues.

In [179]:
mask = adverse_rxn_df.duplicated()
adverse_rxn_df[mask]

,given_name,surname,adverse_reaction


In [180]:
mask = adverse_rxn_df.duplicated(subset=['given_name', 'surname'])
adverse_rxn_df[mask]

,given_name,surname,adverse_reaction


**Observation :** No issues found in `adverse_rxn_df`.

#### **4. describe Method**

The `describe` method of Pandas provides us with the statistical summary of numerical columns in the DataFrame.  
With this, we can find issues in the distribution of columns.

In [181]:
patients_df.describe()

,patient_id,zip_code,weight,height,bmi
count,503.000000,491.000000,503.000000,503.000000,503.000000
mean,252.000000,49084.118126,173.434990,66.634195,27.483897
std,145.347859,30265.807442,33.916741,4.411297,5.276438
min,1.000000,1002.000000,48.800000,27.000000,17.100000
25%,126.500000,21920.500000,149.300000,63.000000,23.300000
50%,252.000000,48057.000000,175.300000,67.000000,27.200000
75%,377.500000,75679.000000,199.500000,70.000000,31.750000
max,503.000000,99701.000000,255.900000,79.000000,37.700000


In [182]:
suspect = patients_df[patients_df['weight'] == patients_df['weight'].min()].copy()
suspect['calculated_bmi'] = suspect['weight'] / (suspect['height'] ** 2) * 703
suspect[['given_name', 'surname', 'weight', 'height', 'bmi', 'calculated_bmi']]

,given_name,surname,weight,height,bmi,calculated_bmi
210,Camilla,Zaitseva,48.8,63,19.1,8.643588


**Observations**  
1. Column `zip_code` contains a minimum value of 4-digits, which is wrong.
2. Column `weight` contains a minimum value of **48.8 Pounds**, equivalent to **22.13531 Kg**, which is suspicious.  
   On looking for that person details about `weight` and `height`, it is found that the weight information is incorrect.  
   This is verified by calculating the `bmi` of that person on the basis of given weight & height, which mismatches with the `bmi` in the data.  
   **Note :** This raises a concern that there is possibility that most of the `bmi` are calculated wrong.

So, now lets check for the `bmi` calculations issue.

In [183]:
def calc_bmi(df):
    '''Calculate BMI.'''
        
    return round(df['weight'] / (df['height'] ** 2) * 703, 1)

In [184]:
df = patients_df.copy()
df['calculated_bmi'] = calc_bmi(df)
incorrect_bmi_mask = df['bmi'] != df['calculated_bmi']
df[incorrect_bmi_mask]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi,calculated_bmi
4,5,male,Tim,Neudorf,1428 Turkey Pen Lane,Dothan,AL,36303.0,United States,334-515-7487TimNeudorf@cuvox.de,2/18/1928,192.3,27,26.1,185.4
210,211,female,Camilla,Zaitseva,4689 Briarhill Lane,Wooster,OH,44691.0,United States,330-202-2145CamillaZaitseva@superrito.com,11/26/1938,48.8,63,19.1,8.6


**Observation :** There are two person (patient_id = 5, 211) with unusal weight & height values. So, lets document it.

Now, lets analyze the `treatments_df` using `describe` method.

In [185]:
treatments_df.describe()

,hba1c_start,hba1c_end,hba1c_change
count,280.000000,280.000000,171.000000
mean,7.985929,7.589286,0.546023
std,0.568638,0.569672,0.279555
min,7.500000,7.010000,0.200000
25%,7.660000,7.270000,0.340000
50%,7.800000,7.420000,0.380000
75%,7.970000,7.570000,0.920000
max,9.950000,9.580000,0.990000


Since, `hba1c_change` is calculated as $hba1c_{start} - hba1c_{end}$. We can verify if there are any mistakes in calculation of `hba1c_change`.

In [186]:
df = treatments_df.copy()
df['hba1c_calculated'] = (df['hba1c_start'] - df['hba1c_end'])
wrong_hba1c_change = (
    df['hba1c_change'].notnull() &
    ~np.isclose(a=df['hba1c_change'], b=df['hba1c_calculated'],
                rtol=1e-5, atol=1e-8)
)
df[wrong_hba1c_change]

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change,hba1c_calculated
1,elliot,richardson,-,40u - 45u,7.56,7.09,0.97,0.47
13,gregor,bole,-,47u - 45u,7.61,7.16,0.95,0.45
17,gina,cain,-,36u - 36u,7.88,7.40,0.98,0.48
20,mile,stanić,-,47u - 48u,7.66,7.24,0.92,0.42
25,benoît,bonami,-,44u - 43u,9.82,9.40,0.92,0.42
26,suhaim,rahal,-,49u - 47u,7.94,7.50,0.94,0.44
27,mizuki,iwata,-,45u - 46u,7.70,7.23,0.97,0.47
32,laura,ehrlichmann,-,43u - 40u,7.95,7.46,0.99,0.49
35,csaba,sági,-,35u - 29u,7.88,7.48,0.90,0.40
40,ásta,grímsdóttir,-,29u - 30u,7.62,7.16,0.96,0.46


**Observation :** It can be seen above that **56** values of `hba1c_change` are miscalculated. So, lets document it.

### 🎯 **Data Quality Dimensions**

*DataCamp Cheatsheet* on Data Quality Dimensions : [Link](https://media.datacamp.com/legacy/image/upload/v1678808802/Data_Quality_Dimensions_yebbuu.pdf)

There are mainly six types of data quality dimensions that are used to assess the quality of the data.  
1. Completeness
2. Uniqueness
3. Validity
4. Accuracy
5. Timeliness
6. Consistency

#### **1. Completeness**  
Completeness refers to the extent to which all required data is present, available, and recorded in the dataset.  

**Example :**  
| Order ID | Product | Price | Delivery Date |
| -------- | ------- | ----- | ------------- |
| 8001     | Laptop  | 60000 | 2024-06-10    |
| 8002     | Phone   | 32000 | *Missing*     |


#### **2. Uniqueness**  
Uniqueness measures the degree to which the records in a dataset are not duplicated.  

**Example :**  

| Customer ID | Name | Email                                 |
| ----------- | ---- | ------------------------------------- |
| 101         | Ali  | ali@gmail.com |
| 101         | Ali  | ali@gmail.com |


#### **3. Validity**  
Validity refers to how well the data follows the defined rules, formats, business constraints, or acceptable ranges.  

**Example :**  
Valid Range for Age is [0, 120].
| Age | Valid? | Reason               |
| --- | ------ | -------------------- |
| 25  | Yes      | Within range         |
| -5  | No      | Negative not allowed |
| 150 | No      | Exceeds max age      |


#### **4. Accuracy**  
Accuracy measures the degree to which data is correct and represents the truth.  
In other words, it refers to how correctly the data represents real world values.  

**Note :**  
1. A value can be *complete, valid, and consistent* yet still be ***inaccurate***.
2. Valid data can still be inaccurate, but invalid data is always inaccurate.  

**Example :** The actual price of a phone is ***1000 USD***. In data, it is stored as ***100 USD***. This is the problem of accuracy.  
Here, it can be seen that ***100 USD*** is *complete, valid, and consistent*, yet it is ***inaccurate*** because it doesn't matches with the actual price of the phone.


#### **5. Timeliness**  
Timeliness refers to how up-to-date, current, and available the data is when it is needed.  

**Example :**  

1. Banks need real-time data.
2. If your account balance updates after 2 days → not timely
3. Even if the numbers are correct, they arrive too late to be useful.

#### **6. Consistency**  
Consistency means that data should not contradict itself across different sources, tables, fields, or systems.  

**Example :**  

**Source A: CRM System**  
Customer Name: Aisha Khan  
Email: aisha@gmail.com  

**Source B: Sales System**  
Customer Name: Ayesha Khan  
Email: aisha_khan@gmail.com  

These contradictions show inconsistency.

### ⚠️ **Severity Order**  

| Priority | Category        | Description                            |
| -------- | --------------- | -------------------------------------- |
| **1**    | Completeness    | Handling Missing Values                |
| **2**    | Uniqueness      | Remove duplicates                      |
| **3**    | Tidiness        | Fix structure/shape of the dataset     |
| **4**    | Accuracy        | Correct factually wrong data           |
| **5**    | Validity        | Fix invalid formats, impossible values |
| **6**    | Consistency     | Resolve contradictions across data     |
| **7**    | Timeliness      | Handle outdated data                   |


Now, lets label all the issues that we have found in the previous assessment steps.  
We will be labelling the issues with the dimensions that we just studied.

### 🎯 **Labelling Issues**

1. **Dirty Issues**
   1. ***Table : Patients***
      1. **Manual**
         1. The value of `given_name` for `patient_id = 9` is *Dsvid* instead of *David*. ***Accuracy***
         2. There are some *missing values* in columns `[address, city, state, zip_code, country, contact]`. ***Completeness***
         3. In the `state` column, values are inconsistent. Some values are *abbreviations*, whereas some are *full forms*. ***Consistency***
         4. In the `zip_code` column, values are inconsistent. Some values are *4-digit* zip code, which is not possible. ***Validity***
      2. **Programmatically**
         1. There are **12** *missing values* in columns `[address, city, state, zip_code, country, contact]`. ***Completeness***
         2. Column `assigned_sex` should have `pd.Categorical` as its data type. ***Validity***
         3. Column `zip_code` should have `int64` as its data type. ***Validity***
         4. Column `birthdate` should have `pd.DateTime` as its data type. ***Validity***
         5. Duplicated Entries by the name **John Doe** with different `patient_id` and some other duplicate entries by following names:  
            1. **Jakob Jakobsen** and **Jake Jakobsen** are same. ***Uniqueness***
            2. **Patrick Gersten** and **Pat Gersten** are same. ***Uniqueness***
            3. **Sandra Taylor** and **Sandy Taylor** are same.  ***Uniqueness***
         6. Unusual value of `weight` & `height` were seen in `patient_id` 5 & 211. ***Accuracy***
   2. ***Table : Treatments & Treatments Cut***
      1. **Manual** 
         1. The `given_name` and `surname` has values in lower case, which is not consistent with the *Patients Table*. ***Consistency***
         2. Remove `u` from the values of `auralin` and `novodra`, since the units of all values are same. ***Validity***
         3. Hyphens (-) in `auralin` and `novodra` columns, are actually `nan` values, but ain't getting treated like that. ***Validity***
         4. Column `hba1c_change` contains some *missing values*. ***Completeness***
      2. **Programmatically**
         1. Column `hba1c_change` contains **109** *missing values* in `treatments_df` and **28** in `treatments_cut_df`. ***Completeness***
         2. While splitting `auralin`, assign proper data types to `auralin_start` and `auralin_end`. ***Validity***
         3. While splitting `novodra`, assign proper data types to `novodra_start` and `novodra_end`. ***Validity***
         4. A duplicate entry exists by the name **joseph day** in `treatments_df`. ***Uniqueness***
         5. **56** values in `hba1c_change` column are miscalculated. ***Accuracy***
   3. ***Table : Adverse Reactions***
      1. The `given_name` and `surname` has values in lower case, which is not consistent with the *Patients Table*. ***Consistency***
2. **Messy Issues**
   1. ***Table : Patients***
      1. The `contact` column contains both *phone number* & *email*. 
   2. ***Table : Treatments & Treatments Cut***
      1. **Manual**
         1. The column `auralin` should get splitted into two columns `auralin_start` and `auralin_end`.
         2. The column `novodra` should get splitted into two columns `novodra_start` and `novodra_end`.
         3. Combine `treatments` and `treatments_cut` tables into one.
   3. ***Table : Adverse Reactions***
      1. This table is capturing information about one variable only, so it should not exist independently.

### 🎯 **Data Cleaning**  

#### **🧹 Cleaning Patients Data**

Now, lets start with data cleaning. We will start with `patients_df`.  
First, we will be resolving the *completeness* issues.  

Before starting with data cleaning, we should always work on the copy of our dataset.  So, lets create copies of all DataFrames.

In [187]:
patients_df = patients_df.copy()
treatments_df = treatments_df.copy()
treatments_cut_df = treatments_cut_df.copy()
adverse_rxn_df = adverse_rxn_df.copy()

Now, lets handle missing values of our `patients_df`.

In [188]:
mask = patients_df['address'].isnull()
patients_df[mask]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
209,210,female,Lalita,Eldarkhanov,NaN,NaN,NaN,NaN,NaN,NaN,8/14/1950,143.4,62,26.2
219,220,male,Mỹ,Quynh,NaN,NaN,NaN,NaN,NaN,NaN,4/9/1978,237.8,69,35.1
230,231,female,Elisabeth,Knudsen,NaN,NaN,NaN,NaN,NaN,NaN,9/23/1976,165.9,63,29.4
234,235,female,Martina,Tománková,NaN,NaN,NaN,NaN,NaN,NaN,4/7/1936,199.5,65,33.2
242,243,male,John,O'Brian,NaN,NaN,NaN,NaN,NaN,NaN,2/25/1957,205.3,74,26.4
249,250,male,Benjamin,Mehler,NaN,NaN,NaN,NaN,NaN,NaN,10/30/1951,146.5,69,21.6
257,258,male,Jin,Kung,NaN,NaN,NaN,NaN,NaN,NaN,5/17/1995,231.7,69,34.2
264,265,female,Wafiyyah,Asfour,NaN,NaN,NaN,NaN,NaN,NaN,11/3/1989,158.6,63,28.1
269,270,female,Flavia,Fiorentino,NaN,NaN,NaN,NaN,NaN,NaN,10/9/1937,175.2,61,33.1
278,279,female,Generosa,Cabán,NaN,NaN,NaN,NaN,NaN,NaN,12/16/1962,124.3,69,18.4


We will be filling missing values appropriately for each column.  
Here is a list of values that will represent missing values in the following columns:  
1. `address` : No Data.
2. `city` : No Data.
3. `state` : No Data.
4. `zip_code` : 0.
5. `country` : No Data.
6. `contact` : No Data.  

Now, lets fill missing values.

In [189]:
patients_df.fillna(
    {
        'address' : 'No Data',
        'city' : 'No Data',
        'state' : 'No Data',
        'zip_code' : 0,
        'country' : 'No Data',
        'contact' : 'No Data',
    },
    inplace=True
)

Lets verify the missing data imputation.

In [190]:
patients_df.iloc[[209]]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi
209,210,female,Lalita,Eldarkhanov,No Data,No Data,No Data,0.0,No Data,No Data,8/14/1950,143.4,62,26.2


Now, resolving the following *uniqueness* issue.  
- Duplicate entries by following names:  
    1. **Jakob Jakobsen** and **Jake Jakobsen** are same. ***Uniqueness***  
    2. **Patrick Gersten** and **Pat Gersten** are same. ***Uniqueness***  
    3. **Sandra Taylor** and **Sandy Taylor** are same.  ***Uniqueness***  

From the *contact* column, correct names of the individuals can be easily identified using email.  
Fortnately, the very first entry of each duplicate is the correct one. So, we will drop other entries except the first one.

In [191]:
patients_df.drop_duplicates(subset=['surname', 'address'], keep='first', inplace=True)

In [192]:
mask = patients_df.duplicated(subset=['surname', 'address'], keep=False)
patients_df[mask]

,patient_id,assigned_sex,given_name,surname,address,city,state,zip_code,country,contact,birthdate,weight,height,bmi


Now, we will be resolving the *messy* issues in our `patients_df`, which is as follows:  
The `contact` column contains both *phone number* & *email*.   
So, we will be splitting the `contact` column into two different columns `phone` and `email`.  
For that, we will have to ***regex*** to extract phone number and email from the contact column.

In [193]:
import re

# Phone Pattern
PHONE_RE = re.compile(r"""
    (?x)(?:\+?1[\s\-.]*)?(?:\(?\d{3}\)?[\s\-.]*)\d{3}[\s\-.]*\d{4}|(?<!\d)\d{10}(?!\d)
""")

# Email Pattern
EMAIL_RE = re.compile(r"[A-Za-z][A-Za-z0-9._%+\-]*@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}")


In [194]:
def _clean_markdown_mailto(s: str) -> str:
    """Turns [text](mailto:someone@x.com) into text; keeps parentheses used in phone numbers."""
    if s is None:
        return None
    s = str(s)
    s = re.sub(r"\[([^\]]+)\]\(mailto:[^)]+\)", r"\1", s)
    s = s.replace("mailto:", "")
    return s

def _normalize_us_phone(phone_raw: str) -> str:
    if not phone_raw:
        return None
    digits = re.sub(r"\D", "", phone_raw)
    if len(digits) == 11 and digits.startswith("1"):
        digits = digits[1:]
    if len(digits) != 10:
        return phone_raw.strip()
    return f"{digits[0:3]}-{digits[3:6]}-{digits[6:10]}"

def extract_email_phone(contact: str):
    s = _clean_markdown_mailto(contact)
    if s is None or s.strip() == "":
        return pd.Series({"email": None, "phone": None})

    email_match = EMAIL_RE.search(s)
    email = email_match.group(0) if email_match else None

    s_wo_email = s.replace(email, " ") if email else s
    phone_match = PHONE_RE.search(s_wo_email)
    phone_raw = phone_match.group(0) if phone_match else None
    phone = _normalize_us_phone(phone_raw)

    return pd.Series({"phone": phone, "email": email})

Now, lets apply these functions to extract *phone* and *email*.

In [195]:
patients_df[['phone', 'email']] = patients_df['contact'].apply(extract_email_phone)

In [196]:
patients_df[['contact', 'phone', 'email']].head()

,contact,phone,email
0,951-719-9170ZoeWellish@superrito.com,951-719-9170,ZoeWellish@superrito.com
1,PamelaSHill@cuvox.de+1 (217) 569-3204,217-569-3204,PamelaSHill@cuvox.de
2,402-363-6804JaeMDebord@gustr.com,402-363-6804,JaeMDebord@gustr.com
3,PhanBaLiem@jourrapide.com+1 (732) 636-8246,732-636-8246,PhanBaLiem@jourrapide.com
4,334-515-7487TimNeudorf@cuvox.de,334-515-7487,TimNeudorf@cuvox.de


Now, lets remove the `contact` column, as it is not needed.

In [197]:
patients_df.drop(columns=['contact'], inplace=True)

Now, lets start resolving the accuracy issues. The accuracy issues in `patients_df` are as follows:  
1. The value of `given_name` for `patient_id = 9` is *Dsvid* instead of *David*.
2. Unusual value of `weight` & `height` were seen in `patient_id` 5 & 211. ***Accuracy***

Lets resolve the first one manually.

In [198]:
mask = patients_df['patient_id'] == 9
patients_df.loc[mask, 'given_name'] = 'David'

In [199]:
patients_df.loc[mask, 'given_name']

8    David
Name: given_name, dtype: object

Now, lets resolve the second issue.

In [200]:
mask = patients_df['patient_id'].isin([5, 211])
patients_df.loc[mask, ['patient_id', 'weight', 'height', 'bmi']]

,patient_id,weight,height,bmi
4,5,192.3,27,26.1
210,211,48.8,63,19.1


Fixing *height* of patient with ID 5.

In [201]:
from math import sqrt

mask = patients_df['patient_id'] == 5
patients_df.loc[mask, 'height'] = round(sqrt(
    (patients_df.loc[mask, 'weight'] / patients_df.loc[mask, 'bmi'] * 703).iloc[0]), 1)

Fixing *weight* of patient with ID 211.

In [202]:
mask = patients_df['patient_id'] == 211
patients_df.loc[mask, 'weight'] = round(
    (patients_df.loc[mask, 'bmi'] * pow(patients_df.loc[mask, 'height'], 2) / 703).iloc[0], 1)

Assuring changes.

In [203]:
mask = patients_df['patient_id'].isin([5, 211])
patients_df.loc[mask, ['patient_id', 'weight', 'height', 'bmi']]

,patient_id,weight,height,bmi
4,5,192.3,72,26.1
210,211,107.8,63,19.1


Now, lets start resolving *Validity* Issues. Here are those issues:  
1. In the `zip_code` column, values are inconsistent. Some values are *4-digit* zip code, which is not possible. ***Validity***
2. Column `assigned_sex` should have `pd.Categorical` as its data type. ***Validity***
3. Column `zip_code` should have `int64` as its data type. ***Validity***
4. Column `birthdate` should have `pd.DateTime` as its data type. ***Validity***

Starting with the `zip_code` column.  


While assessing the `zip_code` column, I found that the 4-digit zip codes are actually result of a behaviour where zeros in the starting are ignored as the data type of the `zip_code` column is `float`. For example:  
1. 01234 is interpreted as 1234 while reading the CSV file.
2. Similarly, 04765 is interpreted as 4765.  

Now, lets tackle this issue.  

**NOTE :** The point to be noted is that some of the missing values in the `zip_code` column is imputed as `0` when the data type was `float64`. But now, I will be changing the data type to `str`, so representation of missing values will be changed to `No Data`.

In [204]:
def validate_zip_code(code):
    if code == '0':
        return 'No Data'
    elif len(code) < 5:
        return code.zfill(5)
    else:
        return code

In [205]:
patients_df['zip_code'] = patients_df['zip_code'].astype(int).astype(str).apply(validate_zip_code)

Now, assuring changes in `zip_code` column.

In [206]:
patients_df['zip_code'].apply(lambda x : len(x)).value_counts()

zip_code
5    483
7     12
Name: count, dtype: int64

Now, fixing the data type of `assigned_sex` column.

In [207]:
patients_df['assigned_sex'] = patients_df['assigned_sex'].apply(lambda x : 'M' if x == 'male' else 'F').astype('category')

Now, fixing the data type of `birthdate` column.

In [208]:
patients_df['birthdate'] = pd.to_datetime(patients_df['birthdate'])

Now, lets start dealing with the *Consistency* Issues.  

There's only one consistency issue in `patients_df` as follows :  
1. In the `state` column, values are inconsistent. Some values are *abbreviations*, whereas some are *full forms*. ***Consistency***

In [209]:
state_abbreviations = {
    'California': 'CA',
    'Texas': 'TX',
    'New York': 'NY',
    'Massachusetts': 'MA',
    'Pennsylvania': 'PA',
    'Georgia': 'GA',
    'Illinois': 'IL',
    'Ohio': 'OH',
    'Florida': 'FL',
    'Michigan': 'MI',
    'Oklahoma': 'OK',
    'Louisiana': 'LA',
    'New Jersey': 'NJ',
    'Virginia': 'VA',
    'Mississippi': 'MS',
    'Wisconsin': 'WI',
    'Indiana': 'IN',
    'Minnesota': 'MN',
    'Tennessee': 'TN',
    'Alabama': 'AL',
    'North Carolina': 'NC',
    'Kentucky': 'KY',
    'Washington': 'WA',
    'Missouri': 'MO',
    'Idaho': 'ID',
    'Kansas': 'KS',
    'Nevada': 'NV',
    'South Carolina': 'SC',
    'Iowa': 'IA',
    'Connecticut': 'CT',
    'Maine': 'ME',
    'North Dakota': 'ND',
    'Nebraska': 'NE',
    'Rhode Island': 'RI',
    'Arkansas': 'AR',
    'Colorado': 'CO',
    'Arizona': 'AZ',
    'Maryland': 'MD',
    'Delaware': 'DE',
    'West Virginia': 'WV',
    'Oregon': 'OR',
    'South Dakota': 'SD',
    'Montana': 'MT',
    'Vermont': 'VT',
    'District of Columbia': 'DC',
    'Alaska': 'AK',
    'Wyoming': 'WY',
    'New Hampshire': 'NH',
    'New Mexico': 'NM',
    'No data': 'Unknown'
}

Now, we will use the above dictionary of State and their abbreviations to introduce consistency in the `state` column.

In [210]:
patients_df['state'] = patients_df['state'].apply(lambda x : state_abbreviations.get(x, x))

#### **🧹 Cleaning Treatments & Treatments Cut Data**

According to the severity order as mentioned earlier in this notebook, we will first start with resolving *Completeness* Issues.  
Here is a *Completeness* issue that are found previously in the `treatments_df` and `treatments_cut_df`:  
1. Column `hba1c_change` contains **109** *missing values* in `treatments_df` and **28** in `treatments_cut_df`. ***Completeness***

Now, lets resolve it.  
Since, `hba1c_change` is calculated as $$ hba1c_{change} = hba1c_{start} − hba1c_{end} $$  
So, filling out *null* values will be easy for us, as we can now use the above formula.

In [211]:
treatments_df['hba1c_change'] = treatments_df['hba1c_start'] - treatments_df['hba1c_end']

Now, filling the null values in the `hba1c_change` column of `treatments_cut_df` in the same way.

In [212]:
treatments_cut_df['hba1c_change'] = treatments_cut_df['hba1c_start'] - treatments_cut_df['hba1c_end']

Now, resolving the *Uniqueness* issue, which is as follows:  
1. A duplicate entry exists by the name **joseph day** in `treatments_df`. ***Uniqueness***

In [213]:
mask = treatments_df.duplicated(keep=False)
treatments_df[mask]

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change
10,joseph,day,29u - 36u,-,7.7,7.19,0.51
136,joseph,day,29u - 36u,-,7.7,7.19,0.51


Now, deduplicating `treatments_df`.

In [214]:
treatments_df.drop_duplicates(keep='first', inplace=True)

Now, assuring the changes.

In [215]:
mask = treatments_df.duplicated(keep=False)
treatments_df[mask]

,given_name,surname,auralin,novodra,hba1c_start,hba1c_end,hba1c_change


Now, heading towards *Tidiness Issues*, we previously found these messy issues.  
1. The column `auralin` should get splitted into two columns `auralin_start` and `auralin_end`.
2. The column `novodra` should get splitted into two columns `novodra_start` and `novodra_end`.
3. Combine `treatments` and `treatments_cut` tables into one.

To fix the issues with `auralin` and `novodra` column, the first approach that comes in our mind is to split `auralin` into `auralin_start` and `auralin_end`, and `novodra` into `novodra_start` and `novodra_end`, which is fine. But it will still contain `null` values, which isn't optimal.  

So, the optimal approach to deal with this issue is that, we should determine the medication type (either `auralin` or `novodra`). Then, we should have two columns `dosage_start` and `dosage_end` to store the start and end dosage values for the particular medication type.  

So, lets do this.

In [216]:
def determine_dosage(row: pd.Series):
    '''Determines Medication Type and Dosage Values.'''
    
    if(row['auralin'] != '-'):
        row['medication'] = 'auralin'
    else:
        row['medication'] = 'novodra'
        
    start = float(row[row['medication']].split('-')[0].replace('u', '').strip())
    end = float(row[row['medication']].split('-')[1].replace('u', '').strip())
    
    row['dosage_start'] = start
    row['dosage_end'] = end
    
    return row

In [217]:
# Initialize New Columns
treatments_df['medication'] = np.nan
treatments_df['dosage_start'] = np.nan
treatments_df['dosage_end'] = np.nan

# Apply Function
treatments_df = treatments_df.apply(determine_dosage, axis=1)
treatments_cut_df = treatments_cut_df.apply(determine_dosage, axis=1)

# Remove Unwanted Columns
treatments_df.drop(columns=['auralin', 'novodra'], inplace=True)
treatments_cut_df.drop(columns=['auralin', 'novodra'], inplace=True)

Now, lets verify the changes.

In [218]:
treatments_df.head()

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,medication,dosage_start,dosage_end
0,veronika,jindrová,7.63,7.20,0.43,auralin,41.0,48.0
1,elliot,richardson,7.56,7.09,0.47,novodra,40.0,45.0
2,yukitaka,takenaka,7.68,7.25,0.43,novodra,39.0,36.0
3,skye,gormanston,7.97,7.62,0.35,auralin,33.0,36.0
4,alissa,montez,7.78,7.46,0.32,novodra,33.0,29.0


In [219]:
treatments_cut_df.head()

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,medication,dosage_start,dosage_end
0,jožka,resanovič,7.56,7.22,0.34,auralin,22.0,30.0
1,inunnguaq,heilmann,7.85,7.45,0.40,auralin,57.0,67.0
2,alwin,svensson,7.78,7.34,0.44,auralin,36.0,39.0
3,thể,lương,7.64,7.22,0.42,novodra,61.0,64.0
4,amanda,ribeiro,7.85,7.47,0.38,auralin,36.0,44.0


Now, lets combine `treatments_df` and `treatments_cut_df` into one.

In [220]:
combined_treatments_df = pd.concat(objs=[treatments_df, treatments_cut_df], ignore_index=True)

Lets verify the operation.

In [221]:
assert combined_treatments_df.shape[0] == (treatments_df.shape[0] + treatments_cut_df.shape[0])

Now, after *Tidiness Issues*, heading towards *Accuracy Issues*. We found only one such issue which is as follows:  
1. **56** values in `hba1c_change` column are miscalculated. ***Accuracy***  

But the catch here is that, we have already resolved this accuracy issue, while we were resolving the missing values issue in `hba1c_change` column.

Now, lets head towards *Validity Issues*, which are as follows:  
1. Remove `u` from the values of `auralin` and `novodra`, since the units of all values are same. ***Validity***
2. Hyphens (-) in `auralin` and `novodra` columns, are actually `nan` values, but ain't getting treated like that. ***Validity***
3. While splitting `auralin`, assign proper data types to `auralin_start` and `auralin_end`. ***Validity***
4. While splitting `novodra`, assign proper data types to `novodra_start` and `novodra_end`. ***Validity***  

It's pretty interesting to observe that all the above *Validity Issues*, were already resolved while tackling other issues.

Now, lets head towards the *Consistency Issue*, which is as follows:  
1. The `given_name` and `surname` has values in lower case, which is not consistent with the *Patients Table*. ***Consistency***

In [222]:
def capitalize(text: str):
    return text.capitalize()

In [223]:
combined_treatments_df['given_name'] = combined_treatments_df['given_name'].apply(capitalize)
combined_treatments_df['surname'] = combined_treatments_df['surname'].apply(capitalize)

Lets verify the operation.

In [224]:
combined_treatments_df.head()

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,medication,dosage_start,dosage_end
0,Veronika,Jindrová,7.63,7.20,0.43,auralin,41.0,48.0
1,Elliot,Richardson,7.56,7.09,0.47,novodra,40.0,45.0
2,Yukitaka,Takenaka,7.68,7.25,0.43,novodra,39.0,36.0
3,Skye,Gormanston,7.97,7.62,0.35,auralin,33.0,36.0
4,Alissa,Montez,7.78,7.46,0.32,novodra,33.0,29.0


#### **🧹 Cleaning Adverse Reactions Data**

The table `adverse_rxn_df` is capturing just one information about the patients. So, from the point of Optimization, it should not exist independently. So, it's better to combine it with the `combined_treatments_df`.  

But, first, we have to capitalize the `given_name` and `surname`, so that we can easily merge it with `combined_treatments_df`.

So, now resolving the *Consistency Issue*.

In [225]:
adverse_rxn_df['given_name'] = adverse_rxn_df['given_name'].apply(capitalize)
adverse_rxn_df['surname'] = adverse_rxn_df['surname'].apply(capitalize)

Lets verify the operation.

In [226]:
adverse_rxn_df.head()

,given_name,surname,adverse_reaction
0,Berta,Napolitani,injection site discomfort
1,Lena,Baer,hypoglycemia
2,Joseph,Day,hypoglycemia
3,Flavia,Fiorentino,cough
4,Manouck,Wubbels,throat irritation


Now, lets remove the existence of `adverse_rxn_df` by merging it with `combined_treatments_df`.

In [227]:
combined_treatments_df = combined_treatments_df.merge(adverse_rxn_df, how='outer')

Lets verify the operation.

In [228]:
combined_treatments_df.head()

,given_name,surname,hba1c_start,hba1c_end,hba1c_change,medication,dosage_start,dosage_end,adverse_reaction
0,Aaliyah,Rice,7.64,7.33,0.31,novodra,31.0,31.0,NaN
1,Abdul-nur,Isa,7.98,7.53,0.45,novodra,54.0,50.0,hypoglycemia
2,Abel,Yonatan,7.88,7.50,0.38,novodra,38.0,39.0,cough
3,Addolorata,Lombardi,7.75,7.33,0.42,novodra,49.0,46.0,NaN
4,Adib,Ghanem,7.59,7.12,0.47,novodra,46.0,49.0,NaN


### ⚠️ **Data Export**
Now lets export the cleaned DataFrames to CSVs to save our rigorous assessing and cleaning work.
You can also find these files in the [Exports](../Resources//Exports/) Folder.

In [229]:
patients_df.to_csv('../Resources/Exports/patients_cleaned.csv')
combined_treatments_df.to_csv('../Resources/Exports/treatments_cleaned.csv')